### PR TITLE
[WIP] Easy Deploy on Save

### PR DESCRIPTION
**Pull request by allileong**
I'd like to get feedback on my implementation approach for Easy Deploy on Save in case my current approach requires major changes.

### What does this PR do?
* Adds a workspace configuration called `salesforcedx-vscode-core.push-or-deploy-on-save.enabled` that is set to `false` by default
* Works in both scratch orgs and non-scratch orgs

Still left to do:
* Update @salesforce/core package to the latest version (in this PR it has been bumped from 0.18.2 to 0.23.1, but the latest version available in npm is 0.24.8)
* Update typescript in all repo packages
* Add testing
* Test on OS's other than Mac
* Test out the alternative way to batch created/deleted files for a single deploy/push; Right now, whenever I detect a create/delete, I extend the wait time by 1 second. When the wait time of 1 second has passed without detecting any additional creates/deletes, then I initiate a single deploy/push

### What issues does this PR fix or reference?
https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000005wp9BIAQ/view

### CODE REVIEW

File: packages/salesforcedx-vscode-core/src/commands/forceApexTestRun.ts [hunk 1]
```diff
@@ -86,7 +86,7 @@ export class ForceApexTestRunCommandFactory {
   private data: ApexTestQuickPickItem;
   private getCodeCoverage: boolean;
   private builder: SfdxCommandBuilder = new SfdxCommandBuilder();
-  private testRunExecutorCommand: Command;
+  private testRunExecutorCommand: Command | undefined;
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/commands/forceApexTestRun.ts]: This was required as a result of updating Typescript.

File: packages/salesforcedx-vscode-core/package.json [hunk 2]
```diff
@@ -50,7 +50,7 @@
     "mock-spawn": "0.2.6",
     "nyc": "^11.0.2",
     "sinon": "^2.3.6",
-    "typescript": "2.6.2",
+    "typescript": "2.9.1",
```
Comment (vazexqi) [inline packages/salesforcedx-vscode-core/package.json]: @allileong - If we are upgrading TypeScript, we should do it consistently across all modules. Usually when I do that, I do it in a separate commit to let the test run fully.

Is there a reason why we only update TypeScript for this one module?

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+async function createSourceFileWatcher(): Promise<vscode.FileSystemWatcher | null> {
+  if (
+    vscode.workspace.workspaceFolders &&
+    vscode.workspace.workspaceFolders.length > 0
+  ) {
+    const sfdxProjectPath = vscode.workspace.workspaceFolders[0].uri.fsPath;
+    const sfdxProject = await SfdxProject.resolve(sfdxProjectPath);
+    const sfdxProjectJson = (await sfdxProject.resolveProjectConfig()) as SfdxProjectJson;
+    const packageDirectoryPaths = sfdxProjectJson.packageDirectories.map(
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts]: I am using the @salesforce/core api to fetch the packageDirectories info from the sfdx-project.json in the workspace. I then create a fileWatcher to detect any creations/changes/deletions made to files inside of the packageDirectories. This approach is not able to differentiate between which files are and aren't source.

File: packages/salesforcedx-vscode-core/package.json (hunk 2 above)
Comment (allileong) [reply packages/salesforcedx-vscode-core/package.json]: @vazexqi I plan to complete that as I finish this work. I just wanted to open a WIP review to make sure that I'm on the right track with my approach.

File: packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts [hunk 4]
```diff
@@ -97,9 +100,35 @@ export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<
...
+      type: 'CONTINUE',
+      data: { filePath: sourcePaths, type: FileType.Source }
+    };
+  }
+}
+
 const workspaceChecker = new SfdxWorkspaceChecker();
 
-export async function forceSourceDeploy(explorerPath: any) {
+export async function forceSourceDeployMultipleSourcePaths(uris: vscode.Uri[]) {
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts]: This new deploy command is only called when a file within an sfdx packageDirectory is created/changed/deleted so it does not need to use the ManifestOrSourcePathGatherer to determine whether the passed in paths belong to source or manifests.

File: packages/salesforcedx-vscode-core/src/context/workspaceOrgType.ts [hunk 5]
```diff
@@ -10,34 +10,58 @@ import * as vscode from 'vscode';
 
 import { ForceConfigGet } from '@salesforce/salesforcedx-utils-vscode/out/src/cli';
 
-export async function setupWorkspaceOrgType() {
+export enum OrgType {
+  SourceTracked,
+  NonSourceTracked
+}
+
+export async function getWorkspaceOrgType(): Promise<OrgType> {
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/context/workspaceOrgType.ts]: I pulled out this getWorkspaceOrgType method so that I could use it inside of pushOrDeployOnSave.ts

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        );
+      }
+    }
+  } catch (e) {
+    if (e.name === 'NamedOrgNotFound') {
+      notificationService.showErrorMessage(
+        nls.localize('error_fetching_auth_info_text')
+      );
+    } else if (e.name === 'NoDefaultusernameSet') {
+      // Do nothing.
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts]: @ntotten, if no defaultusername is set in the workspace, should I notify the customer that their changes are not being deployed/pushed to any org?

File: packages/salesforcedx-vscode-core/package.json (hunk 2 above)
Comment (allileong) [reply packages/salesforcedx-vscode-core/package.json]: @vazexqi that's a good idea. When I finalize my upgrade, I will do it in a separate commit

File: packages/salesforcedx-vscode-core/test/telemetry/MockContext.ts [hunk 7]
```diff
@@ -29,6 +29,7 @@ class MockMemento implements Memento {
 export class MockContext implements ExtensionContext {
   constructor(mm: boolean) {
     this.globalState = new MockMemento(mm);
+    this.workspaceState = new MockMemento(mm);
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/test/telemetry/MockContext.ts]: Another change resulting from the TS upgrade

File: packages/salesforcedx-vscode-core/src/statuses/taskView.ts [hunk 8]
```diff
@@ -88,8 +88,8 @@ export class TaskViewService implements TreeDataProvider<Task> {
 }
 
 export class Task extends TreeItem {
-  public readonly label: string;
-  public readonly collapsibleState: TreeItemCollapsibleState;
+  public readonly label: string | undefined;
```
Comment (allileong) [inline packages/salesforcedx-vscode-core/src/statuses/taskView.ts]: Another change resulting from the TS upgrade

File: packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts (hunk 6 above)
Comment (ntotten) [reply packages/salesforcedx-vscode-core/src/settings/pushOrDeployOnSave.ts]: I think it would be a good idea. Maybe something in the console? "Deploy on save is enabled, but no org is authorized. No files deployed to org." //cc @ruthemmanuelle
